Implement firework rocket & firework star

IvanCraft623 opened the pull request:
## Introduction
Implement items: `Firework Rocket`, `Firework Star` and entity: `Firework Rocket`.

## Follow-up
- Merge https://github.com/pmmp/Color/pull/47 for a litle code reduction
- Merge https://github.com/pmmp/BedrockProtocol/pull/118 to be able to use the appropriate constant
- Merge https://github.com/pmmp/Language/pull/145 for fireworks death message

Requires translations:

| Name | Value in eng.ini |
| :--: | :---: |
| `death.attack.fireworks` | `{%0} went off with a bang` |

## Tests
https://youtu.be/kHs_nBmLRhs

IvanCraft623 committed:
Implement Firework Rocket & Firework Star

IvanCraft623 committed:
Fix PHPStan

IvanCraft623 committed:
Make PHPStan happy

IvanCraft623 committed:
...

IvanCraft623 committed:
Implement explosion multi-color & firework star custom color

IvanCraft623 committed:
Add TODO comment

IvanCraft623 committed:
Some API improvements

IvanCraft623 committed:
Make PHPStan happy

IvanCraft623 committed:
shut up!

IvanCraft623 committed:
Simplification

dktapps commented on src/data/bedrock/FireworkRocketTypeIdMap.php:
```diff
@@ -0,0 +1,67 @@
+	 * @var int[]
+	 * @phpstan-var array<int, int>
+	 */
+	private array $enumToId;
```
These fields are uninitialized

IvanCraft623 replied:
done

IvanCraft623 committed:
Firework types name parity with vanilla

IvanCraft623 committed:
FireworkRocketTypeIdMap: fixed uninitialized fields

IvanCraft623 committed:
oops

IvanCraft623 committed:
Remove unused var

IvanCraft623 committed:
Call parent::onFirstUpdate()

IvanCraft623 committed:
Merge branch 'next-major' into firework

IvanCraft623 committed:
Merge branch 'next-major' into firework

IvanCraft623 committed:
Implement firework death message

IvanCraft623 committed:
Implement Explosive interface

dktapps commented on src/entity/object/FireworkRocket.php:
```diff
@@ -0,0 +1,181 @@
+		if(!$this->isFlaggedForDespawn()){
+			$this->motion->x *= 1.15;
+			$this->motion->y += 0.04;
+			$this->motion->z *= 1.15;
```
This should use addMotion instead of mutating the vector directly

dktapps commented on src/entity/object/FireworkRocket.php:
```diff
@@ -0,0 +1,181 @@
+		parent::syncNetworkData($properties);
+
+		//TODO: Use the appropriate constant
+		$properties->setCompoundTag(EntityMetadataProperties::MINECART_DISPLAY_BLOCK, new CacheableNbt($this->item->getNamedTag()));
```
This has now been updated upstream

dktapps commented on src/item/FireworkRocketExplosion.php:
```diff
@@ -0,0 +1,216 @@
+			if($color !== null){
+				$colors[] = $color;
+			}else{
+				//TODO: should throw an exception?
```
`SavedDataLoadingException` should be thrown here

dktapps commented on src/item/FireworkRocketExplosion.php:
```diff
@@ -0,0 +1,216 @@
+	/**
+	 * @param DyeColor[] $colors
+	 */
+	protected function encodeColors(array $colors) : string{
```
This can be static, and should probably be placed near `decodeColors`

dktapps commented on src/item/FireworkRocketExplosion.php:
```diff
@@ -0,0 +1,216 @@
+	}
+
+	/**
+	 * Returns a the mix of all colors.
```
```suggestion
	 * Returns the mix of all colors.
```

dktapps commented on src/item/FireworkRocketType.php:
```diff
@@ -0,0 +1,71 @@
+			new self("large_ball", fn() => new FireworkLargeExplosionSound()),
+			new self("star", fn() => new FireworkExplosionSound()),
+			new self("creeper", fn() => new FireworkExplosionSound()),
+			new self("burst", fn() => new FireworkExplosionSound()),
```
Not sure what the point of using closures is here?

dktapps commented on src/item/FireworkStar.php:
```diff
@@ -0,0 +1,107 @@
+			$customColor->getR() !== $color->getR() ||
+			$customColor->getG() !== $color->getG() ||
+			$customColor->getB() !== $color->getB()
+		){ //check that $customColor is actually custom.
```
You can compare the toRGBA() value instead of doing this

IvanCraft623 replied:
Better use the new `Color::equals()` method :P

dktapps commented on src/item/FireworkStar.php:
```diff
@@ -0,0 +1,107 @@
+		parent::deserializeCompoundTag($tag);
+
+		$explosionTag = $tag->getTag(self::TAG_EXPLOSION);
+		if (!$explosionTag instanceof CompoundTag) {
```
```suggestion
		if(!$explosionTag instanceof CompoundTag){
```

dktapps commented on src/item/FireworkStar.php:
```diff
@@ -0,0 +1,107 @@
+	}
+
+	/** @return $this */
+	public function clearCustomColor() : self{
```
I think we can forgo this in favour of passing `null` to `setCustomColor`. No point cluttering the API with junk methods.

dktapps commented on src/item/FireworkStar.php:
```diff
@@ -0,0 +1,107 @@
+		return $this->customColor ?? $this->explosion->getColorMix();
+	}
+
+	public function hasCustomColor() : bool{
```
`getCustomColor() !== null` can replace this.

dktapps commented on src/item/FireworkStar.php:
```diff
@@ -0,0 +1,107 @@
+		return $this;
+	}
+
+	public function getColor() : Color{
```
what's the difference between this and the custom colour?

IvanCraft623 replied:
By default the firework star gets the color resulting from the mixture of the colors that the explosion contains, so setting a `customColor` you can choose the color of the item regardless of the color of the explosions.

So, `getColor()` returns the color that is being displayed

This is similar behavior to` Item::getName()` and `Item::getCustomName()`

dktapps replied:
ok, so what's the purpose of this? why can't we just use the colour mix exclusively?

IvanCraft623 replied:
Why not use this feature that is allowing mojang? What if users want to use a custom color but don't want to change the explosion?

dktapps replied:
This isn't answering my question. What is the purpose of this? There's no documentation anywhere.

dktapps commented on src/world/sound/FireworkTwinkleSound.php:
```diff
@@ -0,0 +1,35 @@
+use pocketmine\network\mcpe\protocol\LevelSoundEventPacket;
+use pocketmine\network\mcpe\protocol\types\LevelSoundEvent;
+
+class FireworkTwinkleSound implements Sound{
```
A "twinkle" is a visual effect. It doesn't make me think of a sound.

dktapps commented on src/item/FireworkRocketExplosion.php:
```diff
@@ -0,0 +1,216 @@
+	}
+
+	/**
+	 * Returns the main color, this defines stuff like meta.
```
What's the difference between "main colour" and the other colours? This documentation is really insufficient.

dktapps commented on src/entity/animation/FireworkParticlesAnimation.php:
```diff
@@ -0,0 +1,39 @@
+
+final class FireworkParticlesAnimation implements Animation{
+
+	public function __construct(private FireworkRocket $entity){}
```
```suggestion
	public function __construct(
		private FireworkRocket $entity
	){}
```
This is the typical way we use promoted properties, as it makes it easier to see them.

dktapps commented on src/data/bedrock/item/ItemSerializerDeserializerRegistrar.php:
```diff
@@ -468,6 +470,14 @@ function(Banner $item, int $meta) : void{
+			Ids::FIREWORK_STAR,
+			Items::FIREWORK_STAR(),
+			function(FireworkStar $item, int $meta) : void{
+				$item->getExplosion()->setColors([DyeColorIdMap::getInstance()->fromInvertedId($meta) ?? throw new ItemTypeDeserializeException("Unknown firework star meta $meta")]);
```
what happens if the item has colours in the NBT?

IvanCraft623 replied:
the color in the meta is actually irrelevant, it is always overwritten with the colors in the NBT. In fact we could silently ignore the meta, its existence would rather seem to be a mojang hack.

dktapps reviewed (changes requested):
The implementation looks mostly OK. I haven't tested it yet.

My main sore point on this PR is the woefully inadequate documentation and ambiguous method names.

IvanCraft623 committed:
Merge branch 'major-next' into firework

IvanCraft623 committed:
Use `addMotion` instead of mutating the vector directly

IvanCraft623 committed:
Use appropriate constant for firework item metadata

IvanCraft623 committed:
Throw `SavedDataLoadingException` on finding an unknown color

IvanCraft623 committed:
Make `encodeColors` static and move it near to `decodeColors`

IvanCraft623 committed:
Fix typo on `getColorMix` doc

Co-authored-by: Dylan T. <odigiman@gmail.com>

IvanCraft623 committed:
Remove useless closures for sound getter

IvanCraft623 committed:
Better use the new `Color::equals()` method :P

IvanCraft623 committed:
Remove `clearCustomColor()` in favour of passing `null` to `setCustomColor()`

IvanCraft623 committed:
Merge branch 'firework' of https://github.com/IvanCraft623/PocketMine-MP into firework

IvanCraft623 committed:
Fix little cs

Co-authored-by: Dylan T. <odigiman@gmail.com>

IvanCraft623 committed:
Apply dylan's CS suggestion

Co-authored-by: Dylan T. <odigiman@gmail.com>

IvanCraft623 committed:
Fix flight duaration range to wrap into a signed byte range

IvanCraft623 committed:
Document flight duration

IvanCraft623 committed:
Make FireworkRocket setters fluent

IvanCraft623 committed:
Make FireworkRocketExplosion immutable

IvanCraft623 committed:
Use lists instead of arrays for PHPStan

IvanCraft623 committed:
Improve FireworkRocketExplosion documentation

IvanCraft623 committed:
Rename getMainColor method

IvanCraft623 committed:
Remove hasCustomColor() method

IvanCraft623 committed:
Improve FireworkStar documentation

IvanCraft623 committed:
Fix typo in TAG_FLIGH_DURATION

IvanCraft623 committed:
Rename `FireworkTwinkleSound` to `FireworkCrackleSound`

IvanCraft623 committed:
Fix PHPStan lints

IvanCraft623 committed:
Remove unused import

IvanCraft623 committed:
Implement default values

dktapps commented on src/item/FireworkRocket.php:
```diff
@@ -40,23 +40,34 @@
 
-	public function setFlightDuration(int $duration) : void{
+	/**
+	 * Sets the flight duration of the firework.
```
This is still confusing. It's not really the duration of flight at all. Rather it's a multiplier that is multiplied by a random value to get the actual flight duration.

IvanCraft623 replied:
I have updated it, is it better for you?

dktapps replied:
I think the functions should still be renamed. Perhaps "flight duration multiplier" would be a better name.

dktapps commented on src/item/FireworkRocketExplosion.php:
```diff
@@ -107,56 +107,39 @@ public function __construct(
 
 	/**
-	 * Returns the mix of all colors.
+	 * Returns the mixure of colors form {@link FireworkRocketExplosion::getColors()})
```
```suggestion
	 * Returns the mixure of colors from {@link FireworkRocketExplosion::getColors()})
```

IvanCraft623 committed:
Fix tipo

Co-authored-by: Dylan T. <odigiman@gmail.com>

IvanCraft623 committed:
Merge branch 'firework' of https://github.com/IvanCraft623/PocketMine-MP into firework

IvanCraft623 committed:
Accept explosions instead of the whole item

For this we have to recreate the NBT format of the item, blame mojang for complicating things.

IvanCraft623 committed:
Update getFlightDuration and setFlightDuration

IvanCraft623 commented:
After doing some tests with Fireworks in vanilla it seems to behave like a projectile and in fact it can also be used as one, but PM's Projectile class adds logic that breaks the firework behavior...

dktapps committed:
Merge branch 'major-next' into firework

dktapps commented:
> After doing some tests with Fireworks in vanilla it seems to behave like a projectile and in fact it can also be used as one, but PM's Projectile class adds logic that breaks the firework behavior...

Can you clarify?

IvanCraft623 committed:
Rename flightDuration => flightDurationMultiplier

IvanCraft623 committed:
Rename $fireworksTag=> $fireworkData

IvanCraft623 commented:
> Can you clarify?

nvm, it seems that it is not exactly a projectile, but it can be launched by crossbows, it is also able of ringing a bell.

IvanCraft623 committed:
Merge branch 'minor-next' into firework

dktapps committed:
Merge branch 'minor-next' into firework

IvanCraft623 commented:
Missing features:
- [ ] Fireworks can ring bells
- [ ] Fireworks can be directed in any direction (can be observed when flying with elytras, or using a dispenser).

JavierLeon9966 commented:
> Missing features:
> - [ ] Fireworks can ring bells
> - [ ] Fireworks can be directed in any direction (can be observed when flying with elytras, or using a dispenser).

Also with crossbows, but it can wait until those are implemented first.

ShockedPlot7560 commented:
A PR doesn't have to be all in one, I think the current features are sufficient, the remaining ones are not critical.

IvanCraft623 committed:
Merge branch 'minor-next' into firework

IvanCraft623 committed:
Don't flagForDespawn inside explode()

IvanCraft623 committed:
Use of native enum case accessor

IvanCraft623 committed:
Merge branch 'minor-next' into firework

IvanCraft623 committed:
Modernize the FireworkRocketType enum

ShockedPlot7560 commented on src/item/FireworkStar.php:
```diff
@@ -0,0 +1,106 @@
+class FireworkStar extends Item{
+
+	protected const TAG_EXPLOSION = "FireworksItem"; //TAG_Compound
+	protected const TAG_CUSTOM_COLOR = "customColor"; //TAG_Int
```
Is it really a `c` at the start and not a `C` ?

IvanCraft623 replied:
yep

IvanCraft623 committed:
Merge branch 'minor-next' into firework

ipad54 commented on src/item/FireworkRocketType.php:
```diff
@@ -0,0 +1,46 @@
+	case BURST;
+
+	public function getExplosionSound() : Sound{
+		return match($this){
+			self::SMALL_BALL => new FireworkExplosionSound(),
+			self::LARGE_BALL => new FireworkLargeExplosionSound(),
+			self::STAR => new FireworkExplosionSound(),
+			self::CREEPER => new FireworkExplosionSound(),
+			self::BURST => new FireworkExplosionSound(),
```
```suggestion
		return match($this){
			self::SMALL_BALL,
			self::STAR,
			self::CREEPER,
			self::BURST => new FireworkExplosionSound(),
			self::LARGE_BALL => new FireworkLargeExplosionSound(),
```

IvanCraft623 committed:
Group some rocket types in a single case

Co-authored-by: ipad54 <63200545+ipad54@users.noreply.github.com>

ipad54 commented:
For posterity: firework rocket entity also has a long metadata property `ATTACHED_ENTITY_ID` with ID 18 (Bedrock also has `MINECART_HAS_DISPLAY` boolean property with the same id, but it's inconsistency in the game itself). This property is used for elytra boosting: the firework rocket will start to follow the attached entity and give it a movement boost if it glides.

We don't have elytras implemented right now so I think we don't have to bother about this and the PR looks finished for me.

dktapps committed:
Merge branch 'minor-next' into firework

dktapps commented on src/entity/object/FireworkRocket.php:
```diff
@@ -0,0 +1,201 @@
+
+	/**
+	 * TODO: The entity should be saved and loaded, but this is not possible.
+	 * @see https://bugs.mojang.com/browse/MCPE-165230
```
Not liking this. We should still save & restore them in PM, regardless of vanilla bugs that don't store them.

dktapps commented on src/entity/object/FireworkRocket.php:
```diff
@@ -0,0 +1,201 @@
+	/**
+	 * Returns maximum number of ticks this will live for.
+	 */
+	public function getLifeTicks() : int{
```
`getMaxAgeTicks()` would be more clear IMO

dktapps commented on src/entity/object/FireworkRocket.php:
```diff
@@ -0,0 +1,201 @@
+	 *
+	 * @return $this
+	 */
+	public function setLifeTicks(int $lifeTicks) : self{
```
ditto here

dktapps commented on src/item/FireworkRocket.php:
```diff
@@ -0,0 +1,154 @@
+	public function onInteractBlock(Player $player, Block $blockReplace, Block $blockClicked, int $face, Vector3 $clickVector, array &$returnedItems) : ItemUseResult{
+		$correction = 0.15;
+		$position = $blockClicked->getPosition()->addVector($clickVector);
+		$position = match($face){
+			Facing::DOWN => $position->add(0, -$correction, 0),
+			Facing::UP => $position->add(0, $correction, 0),
+			Facing::NORTH => $position->add(0, 0, -$correction),
+			Facing::SOUTH => $position->add(0, 0, $correction),
+			Facing::WEST => $position->add(-$correction, 0, 0),
+			Facing::EAST => $position->add($correction, 0, 0),
+			default => throw new AssumptionFailedError("Invalid facing $face")
+		};
```
```suggestion
		$position = $position->add(Vector3::zero()->getSide($face)->multiply($correction));
```

If we allowed `getSide()` to accept float values, this would be even simpler.

dktapps replied:
Related to pmmp/Math#86

dktapps commented on src/item/FireworkRocket.php:
```diff
@@ -0,0 +1,154 @@
+			default => throw new AssumptionFailedError("Invalid facing $face")
+		};
+
+		$randomDuration = (($this->flightDurationMultiplier + 1) * 10) + mt_rand(0, 12);
```
Prefer if these magic numbers were moved to constants.

IvanCraft623 committed:
Merge branch 'minor-next' into firework

ghost commented:
Fireworks haven't come since 2022, someone please fix these 😥

dktapps committed:
Merge branch 'minor-next' into firework

dktapps committed:
Easy fixes

dktapps committed:
Merge branch 'minor-next' into firework

dktapps committed:
Properly mark firework as non-saveable with chunk

kostamax27 commented on src/item/FireworkRocketExplosion.php:
```diff
@@ -0,0 +1,190 @@
+	}
+
+	/**
+	 * @param DyeColor[] $colors
```
It seems that Minecraft allow to set custom colors.

https://github.com/Mojang/bedrock-samples/blob/c32ab2ac8ebdd8b418c03990b30597d6b56bc8c9/resource_pack/texts/en_US.lang#L5121

dktapps replied:
Not too clear how that'd work since the colors are encoded as an array of bytes

dktapps replied:
Never mind I'm tripping - I'm pretty sure custom colours are just combinations of dye colours in different quantities

dktapps committed:
Fixes

dktapps committed:
Fix CS

dktapps committed:
Stop accelerating fireworks after 60 ticks age

dktapps committed:
More consistent terminology, tidy code

dktapps committed:
Bring explosion damage logic in line with vanilla
with Java, anyway

dktapps committed:
formatting tidy

IvanCraft623 commented on src/entity/object/FireworkRocket.php:
```diff
@@ -149,29 +149,36 @@ public function explode() : void{
+				for($i = 0; $i < 2; $i++){
+					$target = $position->add(0, 0.5 * $i * $height, 0);
+					foreach(VoxelRayTrace::betweenPoints($this->location, $target) as $blockPos){
+						if($world->getBlock($blockPos)->calculateIntercept($this->location, $target) !== null){
```
Dont remember exactly but pretty sure that i compared with isSolid bc it covered most cases, for example damage should be done anyways if there is any flowable block

dktapps replied:
The old check didn't line up with Java when I compared. It didn't make much sense to me for the likes of signs to prevent firework damage.

dktapps replied:
isSolid is also deprecated and returns false for flowables so that wouldn't have done what you wanted either
